### PR TITLE
Rewrite Driver cpu timing to estimate real cpu usage

### PR DESCRIPTION
**Pull request by dain**
The idea is to record real CPU usage only when the operation wall time was "large".  Otherwise the CPU time is estimated:
```
  totalCpuTime = measure directly from driver process start to end
  totalAccountedCpuTime = sum(operatorCpuMeasuredTime)
  totalUnaccountedCpuTime = totalCpuTime - totalAccountedCpuTime 
  unaccountedCpuWallTime = sum(operatorWallTimeWhenCpuNotMeasured)
  oppreatorEstimatedCpu = operatorCpuMeasuredTime + totalUnaccountedCpuTime * (unaccountedCpuWallTime / operatorWallTimeWhenCpuNotMeasured)
```
Basically, we determine how much CPU time was not directly measured, and spread it amongst the operators based on the wall time spend when CPU time is not measured.

Credit for this idea goes to @electrum

### CODE REVIEW

**Comment by sopel39** [comment]
What's the motivation behind this change? Will `totalUnaccountedCpuTime` be also reported so that pure Driver overhead can also be tracked?

> (unaccountedCpuWallTime / operatorWallTimeWhenCpuNotMeasured)

Shouldn't it be `(operatorWallTimeWhenCpuNotMeasured / unaccountedCpuWallTime)` (e.g: to be fraction < 1.0).

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java [hunk 1]
```diff
@@ -68,14 +69,7 @@
...
 
-    private final AtomicLong intervalWallStart = new AtomicLong();
-    private final AtomicLong intervalCpuStart = new AtomicLong();
-    private final AtomicLong intervalUserStart = new AtomicLong();
-
-    private final AtomicLong processCalls = new AtomicLong();
-    private final AtomicLong processWallNanos = new AtomicLong();
-    private final AtomicLong processCpuNanos = new AtomicLong();
-    private final AtomicLong processUserNanos = new AtomicLong();
+    private final OperationTiming processTimer = new OperationTiming();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: `overallTiming` or `driverTiming`

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    {
+        requireNonNull(overallTiming, "overallTiming is null");
+        checkState(this.overallTiming == null, "Overall timer already set");
+        this.overallTiming = overallTiming;
+    }
+
+    public void recordOperationComplete(OperationTiming timing)
+    {
+        requireNonNull(timing, "timing is null");
+        InternalTiming internalTiming = timings.computeIfAbsent(timing, InternalTiming::new);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: This one seems to be brittle. How about having a `Map<OperatorId, InternalTiming>`, and having `recordOperationComplete` accept `int operatorId`

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        if (overallTiming == null && timings.isEmpty()) {
+            return;
+        }
+
+        long wallEnd = ticker.read();
+
+        long cpuEnd;
+        long userEnd;
+        // remeasure cpu timings if the current interval is large
+        if (wallEnd - intervalWallStart < minNanosForCpuTiming) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: I think we should keep overall cpu time to be always precise. It shouldn't hurt, as that's what we have right now.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: If you decided to keep the estimation - i think you should flip the condition.

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        public long getUserMeasuredNanos()
+        {
+            return userMeasuredNanos;
+        }
+
+        public void recordTimings(long totalUnaccountedCpuNanos, long totalUnaccountedUserNanos, long totalUnaccountedCpuWallNanos)
+        {
+            if (cpuUnaccountedWallNanos > 0) {
+                // measured CPU + this timers portion of the unaccounted CPU
+                operationTiming.recordEstimatedCpuTime(cpuMeasuredNanos + totalUnaccountedCpuNanos * (totalUnaccountedCpuWallNanos / cpuUnaccountedWallNanos));
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: It seems that the formula should be next `totalUnaccountedCpuNanos * currentOperatorUnaccountedFraction`, where `currentOperatorUnaccountedFraction = cpuUnaccountedWallNanos / (double) totalUnaccountedCpuWallNanos`

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                // measured USER + this timers portion of the unaccounted USER
+                operationTiming.recordEstimatedUserTime(userMeasuredNanos + totalUnaccountedUserNanos * (totalUnaccountedCpuWallNanos / cpuUnaccountedWallNanos));
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: same

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java [hunk 6]
```diff
@@ -147,24 +141,14 @@ public Session getSession()
...
-        intervalUserStart.set(currentThreadUserTime());
-    }
-
-    public void recordProcessed()
-    {
-        processCalls.incrementAndGet();
-        processWallNanos.getAndAdd(nanosBetween(intervalWallStart.get(), System.nanoTime()));
-        processCpuNanos.getAndAdd(nanosBetween(intervalCpuStart.get(), currentThreadCpuTime()));
-        processUserNanos.getAndAdd(nanosBetween(intervalUserStart.get(), currentThreadUserTime()));
+        operationTimer.setOverallTimer(processTimer);
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: I think this should be in the if statement.

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        wallStart = ticker.read();
+        cpuStart = currentThreadCpuTime();
+        userStart = currentThreadUserTime();
+
+        intervalWallStart = wallStart;
+        intervalCpuStart = cpuStart;
+        intervalUserStart = userStart;
+    }
+
+    public void setOverallTimer(OperationTiming overallTiming)
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: I got rid of this...  Instead, I pass the timing to the end method

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java (hunk 2 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: Each operator has multiple timing objects.  The idea is they create them and use each one for the timings they want.  The `OperationTiming` class in nested in this class, so we can rely on the behavior of equals an hash code.

File: presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        this.overallTiming = overallTiming;
+    }
+
+    public void recordOperationComplete(OperationTiming timing)
+    {
+        requireNonNull(timing, "timing is null");
+        InternalTiming internalTiming = timings.computeIfAbsent(timing, InternalTiming::new);
+
+        long intervalWallEnd = ticker.read();
+        long intervalWallDuration = intervalWallEnd - intervalWallStart;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: Actually i don't know what was wrong with me, and why did i think we need `abs` here. I was worried about the case when `nanoTime` returns negative value. Neither `max()` nor `abs()` won't help for the case of a numeric overflow over `2^63`. Nevermind then. I don't think you need `max`. I'm pretty sure the `System.nanotime()`  shouldn't return the lesser value in a subsequent call.

[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java]: @arhimondr  https://github.com/prestodb/presto/pull/9867#issuecomment-362021800

**Comment by arhimondr** [comment]
Just some findings.

Turns out `currentThreadCpuTime` implementation for Linux is pretty efficient. By default, `currentThreadCpuTime` and `getCurrentThreadUserTime` are implemented by literally reading and parsing `/proc/self/task/<process id>/stat`. That is really inefficient, because if requires textual parsing: http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/87ee5ee27509/src/os/linux/vm/os_linux.cpp#l5410

However, starting from some version of kernel (pretty archaic, like 2.6.<something>), pthread library exposes lightweight CPU time clocks that JVM can use to get the total cpu time for a given process (thread): http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/87ee5ee27509/src/os/linux/vm/os_linux.cpp#l5391

Although getting the `USER` time is still slow, getting the total `SYS(KERNEL) + USER` cpu time is pretty efficient. 

Here are some benchmarks:

https://github.com/arhimondr/presto/blob/bechamark-cpu-counters/presto-array/src/test/java/com/facebook/presto/array/BenchmarkCPUCounters.java

```
============================
LINUX, 4.15.0-32, BARE METAL
============================

# Run complete. Total time: 00:04:03

Benchmark                      Mode  Cnt     Score     Error  Units
BenchmarkCPUCounters.cpuTime   avgt   20   285.009 ±   2.446  ns/op
BenchmarkCPUCounters.nanoTime  avgt   20    17.281 ±   0.022  ns/op
BenchmarkCPUCounters.userTime  avgt   20  4928.495 ± 211.670  ns/op
```

```
=================
LINUX, 4.11.3-67, DEV VM
=================

# Run complete. Total time: 00:04:02

Benchmark                      Mode  Cnt     Score     Error  Units
BenchmarkCPUCounters.cpuTime   avgt   20   273.611 ±  20.021  ns/op
BenchmarkCPUCounters.nanoTime  avgt   20    29.617 ±   1.540  ns/op
BenchmarkCPUCounters.userTime  avgt   20  8135.688 ± 537.767  ns/op
```

And just out of curiosity: 
(i didn't check how is it implemented)

```
=============
MAC OS High Sierra
=============

# Run complete. Total time: 00:04:02

Benchmark                      Mode  Cnt     Score    Error  Units
BenchmarkCPUCounters.cpuTime   avgt   20  1686.670 ± 42.182  ns/op
BenchmarkCPUCounters.nanoTime  avgt   20    35.061 ±  1.503  ns/op
BenchmarkCPUCounters.userTime  avgt   20  1673.923 ± 29.422  ns/op
```

Getting the CPU time on Linux is still slower by the order of magnitude (~15x) than getting the nanotime, but the performance is pretty reasonable. ~300ns per invocation. What really kills the performance - is getting the user cpu time. It is 400x time slower then getting the nanotime. 

What is even more interesting, with increased concurrency getting the user CPU time is getting even slower, when the timings for getting the nanotime and the overall cpu time remain almost the same:

```
===========================
LINUX, 4.15.0-32, BARE METAL, 4 Threads
===========================

# Run complete. Total time: 00:04:03

Benchmark                      Mode  Cnt     Score    Error  Units
BenchmarkCPUCounters.cpuTime   avgt   20   291.248 ±  6.676  ns/op
BenchmarkCPUCounters.nanoTime  avgt   20    17.635 ±  0.109  ns/op
BenchmarkCPUCounters.userTime  avgt   20  6972.293 ± 95.957  ns/op
```

```
===========================
LINUX, 4.15.0-32, BARE METAL, 8 Threads
===========================

# Run complete. Total time: 00:04:04

Benchmark                      Mode  Cnt     Score    Error  Units
BenchmarkCPUCounters.cpuTime   avgt   20   315.731 ±  2.769  ns/op
BenchmarkCPUCounters.nanoTime  avgt   20    19.300 ±  0.091  ns/op
BenchmarkCPUCounters.userTime  avgt   20  9667.381 ± 87.259  ns/op
```

```
===========================
LINUX, 4.15.0-32, BARE METAL, 12 Threads
===========================

# Run complete. Total time: 00:04:04

Benchmark                      Mode  Cnt      Score     Error  Units
BenchmarkCPUCounters.cpuTime   avgt   20    362.335 ±   7.515  ns/op
BenchmarkCPUCounters.nanoTime  avgt   20     22.000 ±   0.367  ns/op
BenchmarkCPUCounters.userTime  avgt   20  12479.263 ± 196.111  ns/op
```

Given that getting the CPU time is reasonably efficient, maybe we don't need to do any estimates overall. But instead we can collect only the overall cpu time for the inner operators?

**Comment by dain** [comment]
Superseded by #11408

**Pull request closed without merging**